core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/util/ChecksumUtil.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.util;
+
+import static software.amazon.awssdk.checksums.DefaultChecksumAlgorithm.CRC32;
+import static software.amazon.awssdk.checksums.DefaultChecksumAlgorithm.CRC32C;
+import static software.amazon.awssdk.checksums.DefaultChecksumAlgorithm.MD5;
+import static software.amazon.awssdk.checksums.DefaultChecksumAlgorithm.SHA1;
+import static software.amazon.awssdk.checksums.DefaultChecksumAlgorithm.SHA256;
+
+import java.io.InputStream;
+import java.util.Locale;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.checksums.spi.ChecksumAlgorithm;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.ConstantChecksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.Crc32CChecksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.Crc32Checksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.Md5Checksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.SdkChecksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.Sha1Checksum;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.Sha256Checksum;
+
+@SdkInternalApi
+public final class ChecksumUtil {
+
+    private static final String CONSTANT_CHECKSUM = "CONSTANT";
+
+    private ChecksumUtil() {
+    }
+
+    /**
+     * Get the correct checksum header name based on the checksum-algorithm. This is required to be of the form
+     * {@code x-amz-checksum-*}, where '*' is alphanumeric checksum-algorithm-id in lower-case form. Examples include:
+     * <p>
+     * x-amz-checksum-sha256, x-amz-checksum-sha1, x-amz-checksum-crc32, x-amz-checksum-crc32c, x-amz-checksum-md5
+     * </p>
+     */
+    public static String checksumHeaderName(ChecksumAlgorithm checksumAlgorithm) {
+        return "x-amz-checksum-" + checksumAlgorithm.algorithmId().toLowerCase(Locale.US);
+    }
+
+    /**
+     * Gets the SdkChecksum object based on the given ChecksumAlgorithm.
+     */
+    public static SdkChecksum fromChecksumAlgorithm(ChecksumAlgorithm checksumAlgorithm) {
```

Comment:
We may want to create a map for this to make it easier to maintain and may be a micro performance improvement `Map<ChecksumAlgorithm, Supplier<SdkChecksum>>`

Reply:
We can do this for the most, we can't for ConstantChecksum, since it takes a value as part of construction. Updating for this case.